### PR TITLE
Refactor endpoint resolver for AWS services

### PR DESCRIPTION
**Pull request by sayden**
## What does this PR do?

Custom endpoint resolver in AWS services initialisation was broken after the aws sdk upgrade.
This PR remove the custom endpoint resolver, but for the case of non aws bucket in direct aws s3 input

- Closes [#32888](https://github.com/elastic/beats/issues/32888)

### CODE REVIEW

**Review by tommyers-elastic — APPROVED** [review]
nice - does this resolve the reported issue in SQS ? 

do we have an understanding of the difference between this and the built-in default resolver? i.e. can we switch over to that instead of maintaining this one?

**Comment by sayden** [comment]
This is the less invasive way to update the current library without proper testing. We can use the default in a follow up, once we understand how it works and why AWS introduced a breaking change there.

**Comment by tommyers-elastic** [comment]
> without proper testing

could you expand on this a little? 😛

**Comment by sayden** [comment]
> could you expand on this a little?

There are no tests, at all, that actually checks this across all Beats. I like to understand side effects of changing something that affects 5-6 packages.

Hey, but disagree and commit. Seriously, no prob.

**Comment by ruflin** [comment]
Is there a way that the user can change this manually in config or environment variables without having to change it in the code? I'm asking to see if for users that are hit by this issue, if there is also currently a manual workaround?

**Comment by sayden** [comment]
> Is there a way that the user can change this manually in config or environment variables without having to change it in the code? I'm asking to see if for users that are hit by this issue, if there is also currently a manual workaround?

Good question, I have looked at the code and I don't see a workaround atm. The only way to avoid the issue will be to not use a customized endpoint as far as I can see.

This function has 22 references in other parts of the code: billing, cloudwatch, rds, sqs... so it's complicated to predict side effects.

Looking at the old and new code, this change looks like the missing piece. But I think we should consider to decouple those 22 functions to alleviate the side effects produced by small changes like this in the near future.

This issue comes from a breaking change in the AWS library where they removed the generic resolver and introduced specific resolvers for each service: ec2 resolver, sqs resolver, etc. I have some degree of confidence, after looking at alternatives and other solutions in the AWS SDK, that this should solve the issue. But as I said before, we should look at those specific resolvers because they probably do much more for users than the custom generic solution we have now.

**Comment by aspacca** [comment]
> Hey, but disagree and commit. Seriously, no prob.

I disagreed and committed :)

Given other issues the problem does not seem to be limited to the missing `SigningRegion`, but to the overall behaviour of the endpoint resolver (see https://github.com/elastic/integrations/issues/3990 for example).

My point to the change I made is that I expect the internal endpoint resolver of the packages in the aws sdk to be much more tested than any coverage and manual testing we can do for an hot fix.

**Comment by aspacca** [comment]
after discussing with @sayden we agreed on reducing the surface of the change

we need proper manual testing of different scenarios

File: x-pack/libbeat/common/aws/credentials.go [hunk 1]
```diff
@@ -185,7 +185,7 @@ func EnrichAWSConfigWithEndpoint(endpoint string, serviceName string, regionName
 
 		beatsConfig.EndpointResolverWithOptions = awssdk.EndpointResolverWithOptionsFunc(
 			func(service, region string, options ...interface{}) (awssdk.Endpoint, error) {
-				return awssdk.Endpoint{URL: eurl}, nil
+				return awssdk.Endpoint{URL: eurl, SigningRegion: region}, nil
```
Comment (kaiyan-sheng) [inline x-pack/libbeat/common/aws/credentials.go:188]: Should we add a check `endpoint != "amazonaws.com"` at line 171?

Comment (aspacca) [reply x-pack/libbeat/common/aws/credentials.go:188]: if the workaround is to have a zero length endpoint basically we are bypassing the call to `EnrichAWSConfigWithEndpoint`

given so I'm going to push a commit removing all the calls to `EnrichAWSConfigWithEndpoint` and rely on every service internal endpoint resolver from the sdk.

the only scenario were we don't access aws endpoint should be in the `s3poller`, where we can have non-aws s3 compatible bucket. I will test what's needed there.

Comment (aspacca) [reply x-pack/libbeat/common/aws/credentials.go:188]: it seems this can be fixed asap in integrations with removing the default value at https://github.com/elastic/integrations/blob/main/packages/aws/manifest.yml#L68

and releasing a new package version

Comment (aspacca) [reply x-pack/libbeat/common/aws/credentials.go:188]: https://github.com/elastic/integrations/pull/4103

**Comment by tommyers-elastic** [comment]
thanks for getting deeper into this @sayden and @aspacca.

i think it would be have been best to merge the small fix as originally reviewed, then approach the larger refactor in a new PR. this way the short term fix for SQS could be merged, whilst the larger PR goes through review + test. i have removed my original ✅ for now.

File: CHANGELOG.next.asciidoc [hunk 2]
```diff
@@ -10,12 +10,10 @@ https://github.com/elastic/beats/compare/v8.2.0\...main[Check the HEAD diff]
 
 *Affecting all Beats*
 
-* Upgrade to Go 1.18. Certificates signed with SHA-1 are now rejected. See the Go 1.18 https://tip.golang.org/doc/go1.18#sha1[release notes] for details. {pull}32493[32493]
```
Comment (ruflin) [inline CHANGELOG.next.asciidoc:13]: Is that change on purpose?

Comment (aspacca) [reply CHANGELOG.next.asciidoc:13]: no, thanks!

let add it back

**Comment by legoguy1000** [comment]
I'm looking at it now, but there may be an issue with removing the endpoint config option for metrics/sqs/... is for non regular AWS, to include the China AWS and US gov isolated (not govcloud) environments which all have different domains than `amazonaws.com`

**Comment by tlee-elastic** [comment]
Quick question, will we expose a UI option for setting the `default_region` for fleet managed agents as I dont think there is a way to set this currently?

**Comment by aspacca** [comment]
@tlee-elastic 
> Quick question, will we expose a UI option for setting the `default_region` for fleet managed agents as I dont think there is a way to set this currently?

@legoguy1000 is taking care of it at https://github.com/elastic/integrations/pull/4158 👍
